Add initial stubs for the Navigation API

TingPing opened the pull request:
#### 37b116f08e59b3eb3c75e46d368814178d02f4f8
<pre>
Add initial stubs for the Navigation API
<a href="https://bugs.webkit.org/show_bug.cgi?id=262906">https://bugs.webkit.org/show_bug.cgi?id=262906</a>

Reviewed by Alex Christensen.

This patch adds a runtime feature flag for and the initial bindings
for the Navigation API. It does not implement any logic.

Spec: <a href="https://html.spec.whatwg.org/#navigation-api">https://html.spec.whatwg.org/#navigation-api</a>

* LayoutTests/fast/dom/navigation-type-back-forward-expected.txt:
* LayoutTests/fast/dom/navigation-type-navigate-expected.txt:
* LayoutTests/fast/dom/navigation-type-reload-expected.txt:
* LayoutTests/fast/dom/resources/navigation-type-matches-querystring.html:
* LayoutTests/fast/dom/webtiming-expected.txt:
* LayoutTests/fast/dom/webtiming.html:
* LayoutTests/http/tests/misc/resources/webtiming-cross-origin-and-back2.html:
* LayoutTests/http/tests/misc/resources/webtiming-cross-origin-redirect.html:
* LayoutTests/http/tests/misc/resources/webtiming-one-redirect.html:
* LayoutTests/http/tests/misc/resources/webtiming-two-redirects.html:
* LayoutTests/http/tests/misc/webtiming-cross-origin-and-back1-expected.txt:
* LayoutTests/http/tests/misc/webtiming-cross-origin-redirect-expected.txt:
* LayoutTests/http/tests/misc/webtiming-no-origin-expected.txt:
* LayoutTests/http/tests/misc/webtiming-no-origin.html:
* LayoutTests/http/tests/misc/webtiming-one-redirect-expected.txt:
* LayoutTests/http/tests/misc/webtiming-two-redirects-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::navigation):
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/LocalDOMWindow.idl:
* Source/WebCore/page/NavigateEvent.cpp: Added.
(WebCore::NavigateEvent::NavigateEvent):
(WebCore::NavigateEvent::create):
(WebCore::NavigateEvent::intercept):
(WebCore::NavigateEvent::scroll):
(WebCore::NavigateEvent::eventInterface const):
* Source/WebCore/page/NavigateEvent.h: Added.
* Source/WebCore/page/NavigateEvent.idl: Added.
* Source/WebCore/page/Navigation.cpp: Added.
(WebCore::Navigation::Navigation):
(WebCore::createNewResult):
(WebCore::Navigation::scriptExecutionContext const):
(WebCore::Navigation::eventTargetInterface const):
(WebCore::Navigation::navigate):
(WebCore::Navigation::reload):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
(WebCore::Navigation::updateCurrentEntry):
* Source/WebCore/page/Navigation.h: Added.
* Source/WebCore/page/Navigation.idl: Added.
* Source/WebCore/page/NavigationCurrentEntryChangeEvent.cpp: Added.
(WebCore::NavigationCurrentEntryChangeEvent::NavigationCurrentEntryChangeEvent):
(WebCore::NavigationCurrentEntryChangeEvent::create):
(WebCore::NavigationCurrentEntryChangeEvent::eventInterface const):
* Source/WebCore/page/NavigationCurrentEntryChangeEvent.h: Added.
* Source/WebCore/page/NavigationCurrentEntryChangeEvent.idl: Added.
* Source/WebCore/page/NavigationDestination.cpp: Added.
* Source/WebCore/page/NavigationDestination.h: Added.
* Source/WebCore/page/NavigationDestination.idl: Added.
* Source/WebCore/page/NavigationHistoryEntry.cpp: Added.
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::scriptExecutionContext const):
(WebCore::NavigationHistoryEntry::eventTargetInterface const):
* Source/WebCore/page/NavigationHistoryEntry.h: Added.
* Source/WebCore/page/NavigationHistoryEntry.idl: Added.
* Source/WebCore/page/NavigationInterceptHandler.h: Added.
* Source/WebCore/page/NavigationInterceptHandler.idl: Added.
* Source/WebCore/page/NavigationNavigationType.h: Added.
* Source/WebCore/page/NavigationTransition.cpp: Added.
* Source/WebCore/page/NavigationTransition.h: Added.
* Source/WebCore/page/NavigationTransition.idl: Added.
* Source/WebCore/page/NavigationType.idl: Added.
* Source/WebCore/page/Navigator.cpp:

Canonical link: <a href="https://commits.webkit.org/269727@main">https://commits.webkit.org/269727@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fccc820f9f75f9b00594bbb0683fe78572be80bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23395 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24518 "Built successfully") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21612 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23666 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3053 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23947 "Built successfully") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23636 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1072 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20258 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26159 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21150 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/27299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/20363 "Built successfully and passed tests") | [💥 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21338 "An unexpected error occured. Recent messages:OS: Sonoma (14.0), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21411 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22752 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/863 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/30138 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/820 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20928 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6221 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5582 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1275 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/30090 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1124 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6115 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/page/NavigateEvent.h:
```diff
@@ -0,0 +1,104 @@
+        RefPtr<DOMFormData> formData;
+        String downloadRequest;
+        JSC::JSValue info;
+        bool hasUAVisualTransition;
```
If we are free to choose the order, grouping the booleans may give smaller size, this applies especially to the member vars.

rwlbuis commented on Source/WebCore/page/NavigationDestination.h:
```diff
@@ -0,0 +1,57 @@
+    String m_key;
+    String m_id;
+    uint64_t m_index;
+    bool m_sameDocument;
```
You may want to think how to initialize this. If we can go with a default value, these days I prefer stating that in the header, like for example "bool m_sameDocument { false };

nt1m commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -4736,6 +4736,20 @@ MuteCameraOnMicrophoneInterruptionEnabled:
 
+NavigationAPIEnabled:
+  type: bool
+  status: unstable
```
Should it be `testable` so it's enabled in tests?

achristensen07 commented on LayoutTests/fast/dom/resources/navigation-type-matches-querystring.html:
```diff
@@ -12,8 +12,8 @@
     window.performance = window.performance || {};
-    var navigation = performance.navigation || {};
-    shouldBe("navigation.type", stringify(expectedType));
+    window.performance.navgiation = window.performance.navigation || {};
```
I don't think this change is necessary or desired.

achristensen07 commented on LayoutTests/fast/dom/webtiming.html:
```diff
@@ -28,7 +28,7 @@
     shouldBe("timing.redirectStart", "0");
     shouldBe("timing.redirectEnd", "0");
-    shouldBe("navigation.redirectCount", "0");
+    shouldBe("performance.navigation.redirectCount", "0");
```
line 11 is likely no longer needed after these changes.

achristensen07 commented on LayoutTests/fast/dom/resources/navigation-type-matches-querystring.html:
```diff
@@ -12,8 +12,8 @@
-    var navigation = performance.navigation || {};
-    shouldBe("navigation.type", stringify(expectedType));
+    window.performance.navgiation = window.performance.navigation || {};
+    shouldBe("performance.navigation.type", stringify(expectedType));
```
I was initially confused by these changes, but I like the clarification.

achristensen07 commented on Source/WebCore/page/NavigateEvent.h:
```diff
@@ -0,0 +1,104 @@
+        bool hasUAVisualTransition { false };
+    };
+
+    enum class NavigationFocusReset {
```
: bool

achristensen07 commented on Source/WebCore/page/NavigateEvent.h:
```diff
@@ -0,0 +1,104 @@
+        Manual,
+    };
+
+    enum class NavigationScrollBehavior {
```
: bool

achristensen07 commented on Source/WebCore/page/NavigateEvent.h:
```diff
@@ -0,0 +1,104 @@
+    RefPtr<DOMFormData> m_formData;
+    String m_downloadRequest;
+    JSC::JSValue m_info;
+    bool m_canIntercept;
```
Let's give these default values in the header to prevent uninitialized memory.
{ false }

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -0,0 +1,88 @@
+    return { };
+}
+
+Navigation::~Navigation()
```
"= default;" is apparently the cool new way of doing this.  I'm not sure it's better.

achristensen07 commented on Source/WebCore/page/Navigation.h:
```diff
@@ -0,0 +1,108 @@
+    using RefCounted<Navigation>::ref;
+    using RefCounted<Navigation>::deref;
+
+    enum class HistoryBehavior {
```
: uint8_t

achristensen07 commented on Source/WebCore/page/Navigation.h:
```diff
@@ -0,0 +1,108 @@
+    };
+
+    struct Result {
+        bool todo;
```
It would be nice if we could just add a stub implementation of promises that immediately reject instead of this bool that shouldn't be here.

TingPing replied:
Unfortunately the type of `dictionary { Promise, }` isn't handled at all by the binding generation script so I was working on handling this in a later patch.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -0,0 +1,88 @@
+    return NavigationEventTargetInterfaceType;
+}
+
+Navigation::Result Navigation::navigate(const String& /* url */, NavigateOptions&&)
```
This will eventually return ExceptionOr<Navigation::Result> for if the url parsing fails.  Probably several of the other ones will change too.

This is a fine state for a skeleton to be in.  Doesn't need to change this PR.

achristensen07 commented on Source/WebCore/page/Navigation.h:
```diff
@@ -0,0 +1,108 @@
+
+    RefPtr<NavigationHistoryEntry> m_currentEntry;
+    RefPtr<NavigationTransition> m_transition;
+    bool m_canGoBack;
```
{ false } and m_canGoForward too.

achristensen07 commented on Source/WebCore/page/Navigation.idl:
```diff
@@ -0,0 +1,57 @@
+  "auto",
+  "push",
+  "replace"
+};
```
nit: newline.

achristensen07 commented on Source/WebCore/page/NavigationDestination.h:
```diff
@@ -0,0 +1,57 @@
+class NavigationDestination final : public RefCounted<NavigationDestination>, public ScriptWrappable {
+    WTF_MAKE_ISO_ALLOCATED(NavigationDestination);
+public:
+    String url() const { return m_url; };
```
Can these return a const String& to prevent ref churn?

achristensen07 commented on Source/WebCore/page/NavigationDestination.h:
```diff
@@ -0,0 +1,57 @@
+    String id() const { return m_id; };
+    uint64_t index() const { return m_index; };
+    bool sameDocument() const { return m_sameDocument; };
+    JSC::JSValue getState() const { return m_state; };
```
Would it work if this were a const JSC::JSValue& ?

achristensen07 commented on Source/WebCore/page/NavigationInterceptHandler.idl:
```diff
@@ -0,0 +1,2 @@
+[EnabledBySetting=NavigationAPIEnabled] 
```
This file needs a copyright header.

achristensen07 commented on Source/WebCore/page/NavigationNavigationType.h:
```diff
@@ -0,0 +1,37 @@
+
+namespace WebCore {
+
+enum class NavigationNavigationType {
```
: uint8_t

achristensen07 commented on Source/WebCore/page/NavigationNavigationType.h:
```diff
@@ -0,0 +1,37 @@
+namespace WebCore {
+
+enum class NavigationNavigationType {
+    Push = 0,
```
"= 0" doesn't add or change anything here.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4491,6 +4491,8 @@ void WebPage::updatePreferences(const WebPreferencesStore& store)
         updateThrottleState();
     }
 
+    settings.setNavigationAPIEnabled(store.getBoolValueForKey(WebPreferencesKey::navigationAPIEnabledKey()));
```
This line shouldn't be necessary.  If it is, then something is wrong.

TingPing committed:
Add initial stubs for the Navigation API
https://bugs.webkit.org/show_bug.cgi?id=262906

Reviewed by Alex Christensen.

This patch adds a runtime feature flag for and the initial bindings
for the Navigation API. It does not implement any logic.

Spec: https://html.spec.whatwg.org/#navigation-api

* LayoutTests/fast/dom/navigation-type-back-forward-expected.txt:
* LayoutTests/fast/dom/navigation-type-navigate-expected.txt:
* LayoutTests/fast/dom/navigation-type-reload-expected.txt:
* LayoutTests/fast/dom/resources/navigation-type-matches-querystring.html:
* LayoutTests/fast/dom/webtiming-expected.txt:
* LayoutTests/fast/dom/webtiming.html:
* LayoutTests/http/tests/misc/resources/webtiming-cross-origin-and-back2.html:
* LayoutTests/http/tests/misc/resources/webtiming-cross-origin-redirect.html:
* LayoutTests/http/tests/misc/resources/webtiming-one-redirect.html:
* LayoutTests/http/tests/misc/resources/webtiming-two-redirects.html:
* LayoutTests/http/tests/misc/webtiming-cross-origin-and-back1-expected.txt:
* LayoutTests/http/tests/misc/webtiming-cross-origin-redirect-expected.txt:
* LayoutTests/http/tests/misc/webtiming-no-origin-expected.txt:
* LayoutTests/http/tests/misc/webtiming-no-origin.html:
* LayoutTests/http/tests/misc/webtiming-one-redirect-expected.txt:
* LayoutTests/http/tests/misc/webtiming-two-redirects-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::navigation):
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/LocalDOMWindow.idl:
* Source/WebCore/page/NavigateEvent.cpp: Added.
(WebCore::NavigateEvent::NavigateEvent):
(WebCore::NavigateEvent::create):
(WebCore::NavigateEvent::intercept):
(WebCore::NavigateEvent::scroll):
(WebCore::NavigateEvent::eventInterface const):
* Source/WebCore/page/NavigateEvent.h: Added.
* Source/WebCore/page/NavigateEvent.idl: Added.
* Source/WebCore/page/Navigation.cpp: Added.
(WebCore::Navigation::Navigation):
(WebCore::createNewResult):
(WebCore::Navigation::scriptExecutionContext const):
(WebCore::Navigation::eventTargetInterface const):
(WebCore::Navigation::navigate):
(WebCore::Navigation::reload):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
(WebCore::Navigation::updateCurrentEntry):
* Source/WebCore/page/Navigation.h: Added.
* Source/WebCore/page/Navigation.idl: Added.
* Source/WebCore/page/NavigationCurrentEntryChangeEvent.cpp: Added.
(WebCore::NavigationCurrentEntryChangeEvent::NavigationCurrentEntryChangeEvent):
(WebCore::NavigationCurrentEntryChangeEvent::create):
(WebCore::NavigationCurrentEntryChangeEvent::eventInterface const):
* Source/WebCore/page/NavigationCurrentEntryChangeEvent.h: Added.
* Source/WebCore/page/NavigationCurrentEntryChangeEvent.idl: Added.
* Source/WebCore/page/NavigationDestination.cpp: Added.
* Source/WebCore/page/NavigationDestination.h: Added.
* Source/WebCore/page/NavigationDestination.idl: Added.
* Source/WebCore/page/NavigationHistoryEntry.cpp: Added.
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::scriptExecutionContext const):
(WebCore::NavigationHistoryEntry::eventTargetInterface const):
* Source/WebCore/page/NavigationHistoryEntry.h: Added.
* Source/WebCore/page/NavigationHistoryEntry.idl: Added.
* Source/WebCore/page/NavigationInterceptHandler.h: Added.
* Source/WebCore/page/NavigationInterceptHandler.idl: Added.
* Source/WebCore/page/NavigationNavigationType.h: Added.
* Source/WebCore/page/NavigationTransition.cpp: Added.
* Source/WebCore/page/NavigationTransition.h: Added.
* Source/WebCore/page/NavigationTransition.idl: Added.
* Source/WebCore/page/NavigationType.idl: Added.
* Source/WebCore/page/Navigator.cpp:

Canonical link: https://commits.webkit.org/269727@main
